test_conformance/compiler/test_build_helpers.cpp

```diff
@@ -422,6 +451,134 @@ int test_get_program_info(cl_device_id deviceID, cl_context context, cl_command_
     return 0;
 }
 
+int test_get_linked_program_info_kernel_names(cl_device_id deviceID,
+                                              cl_context context,
+                                              cl_command_queue queue,
+                                              int num_elements)
+{
+    int error = CL_SUCCESS;
+    size_t total_kernels = 0;
+    size_t kernel_names_len = 0;
+
+    clProgramWrapper program_AB = clCreateProgramWithSource(
+        context, 1, &sample_multi_kernel_code_AB_with_makro, nullptr, &error);
+    test_error(error, "clCreateProgramWithSource failed");
+
+    clProgramWrapper program_CD = clCreateProgramWithSource(
+        context, 1, &sample_multi_kernel_code_CD_with_makro, nullptr, &error);
+    test_error(error, "clCreateProgramWithSource failed");
+
+    clProgramWrapper program = nullptr;
+
+    // 1) Compile and link the programs with the preprocessor macro undefined.
+    //    Query CL_PROGRAM_NUM_KERNELS and check that the correct number is
+    //    returned. Query CL_PROGRAM_KERNEL_NAMES and check that the right
+    //    kernel names are returned.
+    {
+        error =
+            clCompileProgram(program_AB, 1, &deviceID, nullptr, 0, 0, 0, 0, 0);
+        test_error(error, "clCompileProgram failed");
+
+        error =
+            clCompileProgram(program_CD, 1, &deviceID, nullptr, 0, 0, 0, 0, 0);
+        test_error(error, "clCompileProgram failed");
+
+        cl_program progs[] = { program_AB, program_CD };
+        program =
+            clLinkProgram(context, 1, &deviceID, "", 2, progs, 0, 0, &error);
+        test_error(error, "clLinkProgram failed");
+
+        error = clGetProgramInfo(program, CL_PROGRAM_NUM_KERNELS,
+                                 sizeof(size_t), &total_kernels, nullptr);
+        test_error(error, "clGetProgramInfo failed");
+
+        test_assert_error(total_kernels == 2,
+                          "Unexpected clGetProgramInfo result");
+
+        error = clGetProgramInfo(program, CL_PROGRAM_KERNEL_NAMES, 0, nullptr,
+                                 &kernel_names_len);
+        test_error(error, "clGetProgramInfo failed");
+
+        const size_t len = kernel_names_len + 1;
+        std::vector<char> kernel_names(len, '\0');
+        error =
+            clGetProgramInfo(program, CL_PROGRAM_KERNEL_NAMES, kernel_names_len,
+                             kernel_names.data(), &kernel_names_len);
```

Comment:
If you're interested, it's possible to read into a `std::string` directly (for C++11 and newer, which we are), vs. reading into a `std::vector` and then creating a `std::string` from it.

Something like:

```suggestion
        std::string kernel_names(kernel_names_len, '\0');
        error =
            clGetProgramInfo(program, CL_PROGRAM_KERNEL_NAMES, kernel_names_len,
                             &kernel_names[0], nullptr);
        kernel_names.pop_back(); // remove the final '\0'
```

Reply:
Thanks. Obviously I can apply your suggestion but some time ago I was asked by one of reviewers here in CTS project to use `std::vector<char>` in such cases due to the concerns around spec requirement for `std::string` to have internally continuous chunk of memory. As far as I learnt later it was cleared up at some point in C++11 or 14 spec, still it seems like `std::vector<char>` delivers full control over underlying buffer. Just let me know to do the change.